util/topsql/stmtstats/stmtstats.go

```diff
@@ -15,59 +15,134 @@
 package stmtstats
 
 import (
+	"fmt"
 	"sync"
+	"time"
 
+	"github.com/pingcap/tidb/util/logutil"
 	"go.uber.org/atomic"
+	"go.uber.org/zap"
 )
 
 var _ StatementObserver = &StatementStats{}
 
+// nowFunc returns the current time; it's overridden in tests.
+var nowFunc = time.Now
+
 // StatementObserver is an abstract interface as a callback to the corresponding
 // position of TiDB's SQL statement execution process. StatementStats implements
 // StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
 // The caller only needs to be responsible for calling different methods at the
 // corresponding locations, without paying attention to implementation details.
 type StatementObserver interface {
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnReceiveCmd should be called on TiDB receive command request.
+	OnReceiveCmd()
+
+	// OnSQLAndPlanDigestFirstReady should be called on statement SQL and plan digest are ready.
+	// This function only use to record the SQL and plan digest, it doesn't matter if you delay calling this function.
+	OnSQLAndPlanDigestFirstReady(sqlDigest, planDigest []byte)
 
-	// OnExecutionFinished should be called after the statement is executed.
-	OnExecutionFinished(sqlDigest, planDigest []byte)
+	// OnExecutionFinished should be called after the statement execution finish.
+	OnExecutionFinished()
 }
 
 // StatementStats is a counter used locally in each session.
 // We can use StatementStats to count data such as "the number of SQL executions",
 // and it is expected that these statistics will eventually be collected and merged
 // in the background.
 type StatementStats struct {
+	seCtx    StatementExecutionContext
+	state    StmtExecState
 	mu       sync.Mutex
 	data     StatementStatsMap
 	finished *atomic.Bool
 }
 
+// StmtExecState is the state of StatementStats
+type StmtExecState int
+
+var (
+	stmtExecStateInitial       StmtExecState = 0
+	stmtExecStateCmdReceived   StmtExecState = 1
+	stmtExecStateDigestIsReady StmtExecState = 2
+	stmtExecStateFinished      StmtExecState = 3
+)
+
+// String implements Stringer interface.
+func (s StmtExecState) String() string {
+	switch s {
+	case stmtExecStateInitial:
+		return "initial"
+	case stmtExecStateCmdReceived:
+		return "cmd_received"
+	case stmtExecStateDigestIsReady:
+		return "digest_is_ready"
+	case stmtExecStateFinished:
+		return "finish"
+	default:
+		return fmt.Sprintf("unknow_%d", int(s))
+	}
```

Comment:
If you want to pretty print it, then how about type it as a string, like `type StmtExecState string`?

Reply:
I prefer use int here, since execute compare `int` is faster than compare `string`.

Reply:
I don't think it is really necessary for this trivial optimization. It only reduce less than 1ns. You will never get real performance improvements by focusing on these non performance critical paths. If you really care about the performance, then reduce your allocations, improve your cache efficiency, reduce atomic contentions.

Reply:
Ok, but I still prefer using `int` here.